core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/prompt/PromptGenerator.kt

```diff
@@ -73,16 +73,16 @@ class PromptGenerator(
      */
     fun generatePromptForLine(
         lineUnderTest: String,
-        method: MethodRepresentation,
-        interestingClassesFromMethod: List<ClassRepresentation>,
+        method: MethodRepresentation?,
+        interestingClassesFromMethod: List<ClassRepresentation>?,
```

Comment:
Deeper you get on the method calling chain, more places there will be where you'll need to replace `T` with `T?`. It is generally better to have as few `T?` params as possible; otherwise, your Kotlin code turns into Java.

I recommend resolving all of these problems inside `PromptManager` before calling methods of `PromptGenerator`:
```kotlin
CodeType.LINE -> {
	val lineNumber = codeType.objectIndex

	// get code of line under test
	val document = psiHelper.getDocumentFromPsiFile()
	val lineStartOffset = document!!.getLineStartOffset(lineNumber - 1)
	val lineEndOffset = document.getLineEndOffset(lineNumber - 1)

	val lineUnderTest = document.getText(TextRange.create(lineStartOffset, lineEndOffset))

	val psiMethod = getPsiMethod(cut, getMethodDescriptor(cut, lineNumber))

	val context = if (psiMethod != null) createMethodRepresentation(psiMethod) else cut!!.fullText
	val interestingClassesFromMethod = if (psiMethod != null) {
		psiHelper.getInterestingPsiClassesWithQualifiedNames(cut, psiMethod)
			?.map(this::createClassRepresentation)
			?.toList()
	}
	else {
		emptyList()
	}

	promptGenerator.generatePromptForLine(
		lineUnderTest,
		context,
		interestingClassesFromMethod,
		testSamplesCode,
	)
}
```
And rename the `method` parameter of `promptGenerator.generatePromptForLine` to `context`.

Reply:
The idea is to minimize the number of places where you need to replace `T`  with `T?`. Check all the places where you did the change; likely you will manage to avoid it by having a simple `if` in the caller's code.